### ISSUE TITLE
Flow: return type of Arrow Function

### ISSUE THREAD

**Issue by DmitrySoshnikov** [babel/babel#1991]
Opening this issue to discuss inconsistent position of a return type in an arrow function, and in its type definition.

```javascript
var foo = (x: number): number => x;
```

This is how currently Flow defines the return type of an arrow. It's a parse error in Babel right now, since Babel parses return types _after_ the arrow:

```javascript
var foo = (x: number) => :number x;
```

This is due (I guess), since return type of a _function **type**_ is put after:

```javascript
var bar = (callback: (x: number) => number) => callback(10);
```

If the return type of the function should be before the arrow symbol, we should fix Babel, although, we may probably move it after the arrow to be consistent with a type definition.

CC @gabelevi

**Comment by gabelevi** [comment]
Honestly, I never even considered 

```JavaScript
var foo = (x: number) => :number x;
```

when I added type annotations to arrow functions. The before-the-arrow return type just seemed closer to the `function` syntax. Like

```JavaScript
var f1 = function(x: number): void    {};
var f2 =         (x: number): void => {};
```

Though I do see how I somewhat arbitrarily decided where to place the arrow. The after the arrow syntax does work a little better if you're using the `x => x` arrow function syntax. Currently we completely disallow type annotations in this form, since it's unclear what 

```JavaScript
var f = x: number => x
``` 

means (is number the return type or param type?). With the after the arrow syntax, 

```JavaScript
x: number => : number x
```

is more clear. Nevertheless, I still like the before-the-arrow syntax. It feels cleaner and more intuitive to me. It feels a little weird to see something like `=> : number x`, with no syntax separating the type from the expression. It could definitely cause trouble if we ever add some sort of space to our type language.

**Comment by sebmck** [comment]
Yeah, the current behaviour isn't deliberate, just an oversight with how the parsing is implemented. I agree that before the arrow makes more sense and is far nicer.

**Comment by DmitrySoshnikov** [comment]
Yeah, wondering, can a function type also has it before arrow somehow (though, what to put in the "body" then?):

```js
var foo = (callback: (x: number): void =>) => callback(10);
```

See the "body" of the callback type is empty. Looks weird, but the return type's position is consistent.

**Comment by gabelevi** [comment]
I'm not really sure if they need to be consistent. Function expressions and function types operate on different kind of things. I think the ideal is

```JavaScript
var f1 = (x, y) => x+y;
var f2 : (number, number) => number;
```

So they use the general syntax, only the function type replaces the patterns and expressions with types. Unfortunately, due to TypeScript compatibility we're compromising on the function type to make it

```JavaScript
var f3 : (x: number, y: number) => number;
```

But eventually I'd really like to make the param names optional (which would deviate from TypeScript, which makes the types optional). That would make `f2` the common way to write function types.

For your suggestion, I think we could make the function type `(x: number): number`, but I think it's less obviously a function type and a little harder to read than `(x: number) => number`.

**Comment by DmitrySoshnikov** [comment]
> we could make the function type `(x: number): number`

Yeah, I meant with the arrow: `(x: number): number =>`, but yeah, the empty "body" after the arrow looks weird as well. Anyways, up to the Flow, although, it's already confused me, and probably can be confusing to other people. For now we can fix Babel's parser, and can reiterate it it's confusing.

**Comment by sebmck** [comment]
There's some syntax ambiguity and I'm having some difficulty trying to get around it. Conditionals like:

```javascript
var foo = bar ? (foo) : number;
```

End with a `)` but then can immediately begin with `:` which means that you can't assume that it's the start of a type annotation.

**Comment by DmitrySoshnikov** [comment]
Is it ambiguity in grammar (need to double-check, if yes, then we should restructure it), or an issue in `parseExprAtom`?

FWIW, Flow parser parses it as `foo` identifier (for consequent), and `number` identifier (for alternate).

```
npm install flow-parser
flowparse 'var foo = bar ? (foo) : number;'
```

**Comment by sebmck** [comment]
You should be able to get around it with some tricky parsing, will have a play around and report back.

**Comment by gabelevi** [comment]
Yeah, arrow functions are super ambiguous. Often you don't know if you're dealing with an arrow function until you see the `=>`. For example, `var x = (a, b, c, d)` could either have a grouped sequence expression on the RHS or it could be the beginning of an arrow function. You really don't know until you see the `=>`.

I used to basically keep parallel parse trees to deal with arrow functions, and then collapse to a single parse tree when I was sure what I was dealing with. This was doable since OCaml was forcing me to properly handle it through expression parsing, but it was a pain and was making it really hard to add annotations to arrow functions. Nowadays I'm using an easier-to-work with rollback scheme. The general idea is this:

1. [`Try.to_parse env try_assignment_but_not_arrow_function`](https://github.com/facebook/flow/blob/70deb27e986d40c10cc85b27b93b984b901fbc72/src/parser/parser_flow.ml#L1091). This fails if any of these are true
  - It doesn't parse as an assignment expression
  - [It parses as an assignment expression but is followed by an `=>`](https://github.com/facebook/flow/blob/70deb27e986d40c10cc85b27b93b984b901fbc72/src/parser/parser_flow.ml#L1067)
  - [It parses as an assignment expression but is followed by a `:`](https://github.com/facebook/flow/blob/70deb27e986d40c10cc85b27b93b984b901fbc72/src/parser/parser_flow.ml#L1068)
  - [It parses an assignment expression `async` and is followed by an identifier](https://github.com/facebook/flow/blob/70deb27e986d40c10cc85b27b93b984b901fbc72/src/parser/parser_flow.ml#L1072)
2. If that failed then rollback and [`Try.to_parse env try_arrow_function`](https://github.com/facebook/flow/blob/70deb27e986d40c10cc85b27b93b984b901fbc72/src/parser/parser_flow.ml#L1094). This fails only if it fails to parse it as an arrow function
3. If that failed then rollback and just go with [`assignment_but_not_arrow_function`](https://github.com/facebook/flow/blob/70deb27e986d40c10cc85b27b93b984b901fbc72/src/parser/parser_flow.ml#L1101)

It is structured this way for performance reasons. Almost all assignment expressions should hit step 1 and be done. Arrow functions are much more rare, and will go through to step 2. Malformed expressions and pathological cases will hit step 3. The example you posted is a pathological case.

```JavaScript
var foo = bar ? (foo) : number;
```

So we want to parse an assignment expression after the `?`. So first in step 1 we parse `(foo)` as a valid assignment expression. However we see that it's followed by a `:` so we rollback. Now in step 2 we parse `(foo): number` but don't see a `=>` and fail, so we rollback. Finally, in step 3 we parse `(foo)` and finish successfully.

Hope this explanation of how we parse it helps! I haven't dug into acorn yet so I'm not sure how much it helps.

**Comment by DmitrySoshnikov** [comment]
Yeah, Flow parses an Arrow function is this case for consequent:

```
flowparse 'var foo = bar ? (foo) : number => {} : baz;'
```

The [ConditionalExpression](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-conditional-operator) has the `AssignmentExpression` for the consequent part, and `ArrowFunction` is an `AssignmentExpression`, so it seems should be parseable.

**Comment by DmitrySoshnikov** [comment]
@gabelevi, I see, thanks for the clarifications. The backtracking is slow in general, though seems handles this issue. BTW, would be nice to have an actual grammar written (somewhere in the Flow repo), that people who need to implement Flow plugin in their parsers won't have issue with interpretation. Will the grammar be just normally covered with CoverGrammar? Also, will auto-generated (LR) parsers be able to parse it?

**Comment by gabelevi** [comment]
So the parameter type annotations should be automatically fine with the grammar, since they're basically an extension to `StrictFormalParameters`. The return type annotation I think could be handled with adding a case to `ArrowParameters` to include something like

`CoverParenthesizedExpressionAndArrowParameterList : TypeAnnotation`

So yeah, I haven't thought it through too far, but I don't think type annotations are making this significantly harder to parse...I think it's just tricky due to the spec being like [hey there parser, if you recognize an arrow function that uses `CoverParenthesizedExpressionAndArrowParameterList` then use this supplemental syntax](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-arrow-function-definitions).

**Comment by sebmck** [comment]
Woo, thanks for the awesome explanation @gabelevi! afacd627388dca3b04babd9b5a67d655c88d035d parses everything correctly by backtracking although it's a little rough and I'm a bit worried about performance. FWIW, I've also recently forked Acorn with [babylon](https://github.com/babel/babel/tree/master/packages/babylon) which will hopefully allow us to smoothen out the rough edges and make extensive modifications to better support these kinds of operations.

**Comment by DmitrySoshnikov** [comment]
Cool, thanks for the clarifications guys!

**Comment by sebmck** [comment]
Released as of 5.7.1, thanks!
